quickwit-search/src/retry/mod.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+pub mod search;
+pub mod search_stream;
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use itertools::Itertools;
+use quickwit_proto::{FetchDocsRequest, LeafSearchRequest, LeafSearchStreamRequest};
+
+use crate::client_pool::Job;
+use crate::{ClientPool, SearchClientPool, SearchServiceClient};
+
+/// A retry policy to evaluate if a request should be retried.
+pub trait RetryPolicy<Request, Response, Error>: Sized
+where Request: RequestOnSplit + Clone
+{
+    /// Checks the policy if a certain request should be retried.
```

Comment:
I was surprised by the fact that the prototype suggests that we might retry and `Ok(&Response)`.

I assume this is because potentially, the `Ok(&Response)` might contain a partial set of result?

Reply:
exactly, we want to build a new request on failing splits